### PR TITLE
Make LIKE predicate index aware

### PR DESCRIPTION
**Pull request by alex-dukhno**
Closes https://github.com/hazelcast/hazelcast/issues/14561

### CODE REVIEW

**Comment by taburet** [comment]
Regarding `ILIKE` support, looks like we can use the same trick, but the lower bound should be uppercased while the upper bound should be lowercased.

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java [hunk 1]
```diff
@@ -43,6 +48,28 @@ public LikePredicate(String attributeName, String expression) {
         this.expression = expression;
     }
 
+    @Override
+    public Set<QueryableEntry> filter(QueryContext queryContext) {
+        Index index = queryContext.getIndex(attributeName);
+        if (indexPrefix == null) {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:54]: Caching of the index prefix doesn't make much sense here: `filter` is usually called only once during the predicate lifetime.

Comment (alex-dukhno) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java]: addressed in https://github.com/hazelcast/hazelcast/pull/18289/commits/7e2695ce41ad1208994dafd70be01b14118a792e

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java [hunk 2]
```diff
@@ -43,6 +48,28 @@ public LikePredicate(String attributeName, String expression) {
...
+            indexPrefix = expression.substring(0, expression.length() - 1);
+        }
+        return index.getRecords(indexPrefix, true, indexPrefix + "\uFFFF", false);
+    }
+
+    @Override
+    public boolean isIndexed(QueryContext queryContext) {
+        Index index = queryContext.getIndex(attributeName);
+        if (index != null && index.isOrdered()) {
+            int firstEntry = expression.indexOf('%');
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:61]: As far as I understand, there should be exactly one instance of `%` in the string exactly at its end. There should be no `_` instances anywhere in the string. There is also a possibility of escaping for `%` and `_`, see `Predicates.like` javadoc for more details on the syntax.

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java [hunk 3]
```diff
@@ -43,6 +47,38 @@ public LikePredicate(String attributeName, String expression) {
...
+    @Override
+    public Set<QueryableEntry> filter(QueryContext queryContext) {
+        Index index = queryContext.getIndex(attributeName);
+        String indexPrefix = expression.substring(0, expression.length() - 1);
+        return index.getRecords(indexPrefix, true, indexPrefix + "\uFFFF", false);
+    }
+
+    @Override
+    public boolean isIndexed(QueryContext queryContext) {
+        Index index = queryContext.getIndex(attributeName);
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:59]: It's better to use `queryContext .matchIndex(attributeName, PREFER_ORDERED)` here to make sure we are preferring ordered indexes over the unordered ones (there might be multiple indexes for the same attribute).

Comment (alex-dukhno) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java]: addressed in https://github.com/hazelcast/hazelcast/pull/18289/commits/3ff9fab612cdc376e0ab4300bce4650637188820

**Comment by taburet** [comment]
Looks like there are no tests targeting indexed `LIKE` queries specifically to make sure the queries are really returning only the entries matching the prefix.

**Review by taburet — APPROVED** [review]

**Comment by alex-dukhno** [comment]
Test failure is not related and filled in https://github.com/hazelcast/hazelcast/issues/18323

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java [hunk 4]
```diff
@@ -43,6 +47,38 @@ public LikePredicate(String attributeName, String expression) {
...
+        return index != null && index.isOrdered() && expressionCanBeUsedAsIndexPrefix();
+    }
+
+    private boolean expressionCanBeUsedAsIndexPrefix() {
+        boolean escape = false;
+        for (int i = 0; i < expression.length(); i++) {
+            char c = expression.charAt(i);
+            if (c == '\\') {
+                escape = true;
+            } else {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:69]: What if I have an expression `"\\%"`. Then the `\` is actually escaping the second `\`? But the algorithm will consider like it is escaping the `%` symbol?

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java [hunk 5]
```diff
@@ -43,6 +47,38 @@ public LikePredicate(String attributeName, String expression) {
         this.expression = expression;
     }
 
+    @Override
+    public Set<QueryableEntry> filter(QueryContext queryContext) {
+        Index index = queryContext.matchIndex(attributeName, QueryContext.IndexMatchHint.PREFER_ORDERED);
+        String indexPrefix = expression.substring(0, expression.length() - 1);
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:53]: As I understand we cut the `%` here. But what if we have trailing whitespace? Is it possible?

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java (hunk 4 above)
Comment (alex-dukhno) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:69]: good catch added tests and fixed

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java (hunk 5 above)
Comment (alex-dukhno) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/LikePredicate.java:53]: if `%` is not the last symbol in the expression it wouldn't be indexed, added tests https://github.com/hazelcast/hazelcast/pull/18289/files#diff-8b3c97074f4277303b021451ae7612b8de726a0aec157799b64d5673798b363aR144

**Review by petrpleshachkov — APPROVED** [review]

**Pull request merged**